llvm/lib/CodeGen/MachineCopyPropagation.cpp

```diff
@@ -971,6 +971,18 @@ void MachineCopyPropagation::ForwardCopyPropagateBlock(MachineBasicBlock &MBB) {
 
     forwardUses(MI);
 
+    // It's possible that the previous transformation has resulted in a no-op
+    // register move (i.e. one where source and destination registers are the
+    // same). If so, delete it.
+    CopyOperands = isCopyInstr(MI, *TII, UseCopyInstr);
+    if (CopyOperands &&
```

Comment:
I think you need to check against reserved registers here.

I'm running through the checks in forwardUses, and it looks like none of the others apply.  Just to confirm, the fact the registers are *equal* means that if there are sub-registers involved they must be the same sub-register right?  

I glanced at the MaybeDeadCopies mechanism which seems related, but I don't think we want to use it here as we don't need to worry about block out liveness, or updating debug info here.  Nor does the deadness of this copy depend on following regmasks.

Reply:
I've added a check for reserved registers.

Yes, my understanding and intent is that this only applies for precise register/subregister matches.

Thanks for taking a look through other mechanisms in the file. I'd also looked for something else to piggyback on and came to the same conclusion that there's not existing logic that's quite right.